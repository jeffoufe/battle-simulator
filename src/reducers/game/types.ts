import {
    PLAYER_ROLL,
    MONSTER_ROLL,
    ALL_ROLLS_DONE,
    RESET_TURN,
    COMPUTE_TURN,
    RESET_GAME
} from './constants';

export interface PlayerRollAction {
    type: typeof PLAYER_ROLL
}


export interface MonsterRollAction {
    type: typeof MONSTER_ROLL,
    payload: { result: number }
}

export interface AllRollsDoneAction {
    type: typeof ALL_ROLLS_DONE,
    payload: { result: number }
}

export interface ResetTurnAction {
    type: typeof RESET_TURN
}

export interface ComputeTurnAction {
    type: typeof COMPUTE_TURN
}

export interface ResetGameAction {
    type: typeof RESET_GAME
}

export type GameReducerAction =
    PlayerRollAction |
    MonsterRollAction |
    AllRollsDoneAction |
    ResetTurnAction |
    ComputeTurnAction |
    ResetGameAction