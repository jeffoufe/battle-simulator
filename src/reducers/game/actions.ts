import { 
    PLAYER_ROLL, 
    RESET_GAME,
    MONSTER_ROLL,
    END_TURN
} from './constants';

export const clickAttack = () => ({
    type: PLAYER_ROLL
})

export const resetGame = () => ({
    type: RESET_GAME
})

export const monsterRoll = (result: number) => ({
    type: MONSTER_ROLL,
    payload: { result }
})

export const endTurn = (result: number) => ({
    type: END_TURN,
    payload: { result }
})