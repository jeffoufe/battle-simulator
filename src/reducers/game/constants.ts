export const PLAYER_ROLL = 'GAME/PLAYER_ROLL';
export const MONSTER_ROLL = 'GAME/MONSTER_ROLL';
export const ALL_ROLLS_DONE = 'GAME/ALL_ROLLS_DONE';
export const RESET_TURN = 'GAME/RESET_TURN';
export const END_TURN = 'GAME/END_TURN';
export const COMPUTE_TURN = 'GAME/COMPUTE_TURN';
export const RESET_GAME = 'GAME/RESET_GAME';