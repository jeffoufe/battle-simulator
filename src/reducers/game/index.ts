import { State } from '..';
import { GameReducerAction } from './types';
import {
  PLAYER_ROLL,
  MONSTER_ROLL,
  ALL_ROLLS_DONE,
  RESET_TURN,
  COMPUTE_TURN,
  RESET_GAME
} from './constants';

export const CHARACTER_INITIAL_STATE = {
  health: 100,
  result: 0,
  isRolling: false
};

export const INITIAL_STATE = {
  isTurnActive: false,
  player: CHARACTER_INITIAL_STATE,
  monster: CHARACTER_INITIAL_STATE
}

export default (state: State = INITIAL_STATE, action: GameReducerAction) => {
  switch (action.type) {
    case PLAYER_ROLL:
      return {
        ...state,
        isTurnActive: true,
        player: {
          ...state.player,
          isRolling: true
        }
      }
    case MONSTER_ROLL:
      return {
        ...state,
        monster: {
          ...state.monster,
          isRolling: true
        },
        player: {
          ...state.player,
          isRolling: false,
          result: action.payload.result
        }
      }
    case ALL_ROLLS_DONE:
      return {
        ...state,
        monster: {
          ...state.monster,
          isRolling: false,
          result: action.payload.result
        }
      }
    case COMPUTE_TURN:
      const damage = state.player.result - state.monster.result;
      const character = damage < 0 ? 'player' : 'monster';
      return {
        ...state,
        [character]: {
          ...state[character],
          health: state[character].health - (Math.abs(damage))
        }
      }
    case RESET_TURN:
      return {
        ...state,
        isTurnActive: false,
        player: {
          ...state.player,
          result: 0
        },
        monster: {
          ...state.monster,
          result: 0
        }
      }
    case RESET_GAME:
      return INITIAL_STATE;
    default:
      return state;
  }
};