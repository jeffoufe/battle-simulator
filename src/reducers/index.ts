export { default as gameReducer } from './game';

interface Character {
    isRolling: boolean,
    health: number,
    result: number
};

export interface State {
    player: Character,
    monster: Character,
    isTurnActive: boolean
}