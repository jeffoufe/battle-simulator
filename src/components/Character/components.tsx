import styled from 'styled-components';
import { breakpoints } from '../../constants';

interface CharacterComponentProps {
    order: number
}

export const CharacterContainer = styled.div`
    display: flex;
    flex-direction: ${(props: CharacterComponentProps) => props.order ? 'column-reverse' : 'column'};

    @media only screen and (min-width: ${breakpoints.tablet}) {
        flex-direction: column-reverse;
    }
`

export const CharacterInfoDice = styled.div`
    width: 100%;
    display: flex;
    flex-direction: ${(props: CharacterComponentProps) => props.order ? 'column-reverse' : 'column'};

    @media only screen and (min-width: ${breakpoints.tablet}) {
        flex-direction: ${(props: CharacterComponentProps) => props.order ? 'row-reverse' : 'row'};;
        align-items: center;
        justify-content: space-around;
    }
`

export const CharacterInfo = styled.div`
    display: flex;
    align-items: center;
    justify-content: space-evenly;

    @media only screen and (min-width: ${breakpoints.tablet}) {
        flex-direction: ${(props: CharacterComponentProps) => props.order ? 'row-reverse' : 'row'};
        width: calc(100% - 70px);
    }
`

export const CharacterImage = styled.img`
    width: 85px;
    height: 85px;

    @media only screen and (min-width: ${breakpoints.tablet}) {
        width: 150px;
        height: 150px;
    }

    @media only screen and (min-width: ${breakpoints.laptop}) {
        width: 200px;
        height: 200px;
    }
`;

export const CharacterName = styled.div`
    text-align: center;
    ${(props: CharacterComponentProps) => props.order ? 'margin-top: 10px;' : 'margin-bottom: 10px;'}

    @media only screen and (min-width: ${breakpoints.tablet}) {
        margin-top: 30px;
        font-size: 2em;
    }
`

export const CharacterDice = styled.div`
    ${(props: CharacterComponentProps) => props.order ? 'margin-bottom' : 'margin-top'}: 10px;
    > .dice {
        text-align: center;
    }

    @media only screen and (min-width: ${breakpoints.tablet}) {
        > .dice {
            display: flex;
            flex-direction: column;

            .die-container {
                display: block !important;
                margin: 20px !important;
            }
        }
    }
`