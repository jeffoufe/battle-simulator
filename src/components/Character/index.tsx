import React, { useRef, useEffect, MutableRefObject } from 'react';
import CharacterHealth from '../CharacterHealth';
import ReactDice from 'react-dice-complete';
import 'react-dice-complete/dist/react-dice-complete.css';
import { useDispatch, useSelector } from 'react-redux';
import { MONSTER_ROLL, END_TURN } from '../../reducers/game/constants';

import { 
    CharacterImage, 
    CharacterContainer, 
    CharacterInfoDice,
    CharacterInfo,
    CharacterDice,
    CharacterName
} from './components';

interface CharacterProps {
    image: string,
    order: number,
}

interface DiceRef extends MutableRefObject<null> {
    rollAll: () => void
}


const Character = ({ image, order }: CharacterProps) => {
    const dispatch = useDispatch();
    const { health, isRolling } = useSelector((state: any) => state[order ? 'monster' : 'player'])
    let diceRef = useRef<DiceRef>(null);

    useEffect(() => {
        if (isRolling) {
            diceRef.rollAll();
        }
    }, [isRolling])

    const onRollDone = (result: number) => {
        if (!order) {
            dispatch({
                type: MONSTER_ROLL,
                payload: { result }
            })
        } else {
            dispatch({
                type: END_TURN,
                payload: { result }
            })
        }
    }

    return (
        <CharacterContainer order={order}>
            <CharacterName order={order}>
                {order ? 'Monster' : 'Player'}
            </CharacterName>
            <CharacterInfoDice order={order}>
                <CharacterInfo order={order}>
                    <CharacterImage src={image} />
                    <CharacterHealth health={health} />
                </CharacterInfo>
                <CharacterDice order={order}>
                    <ReactDice
                        numDice={2}
                        faceColor='#FFFFFF'
                        dotColor='#000000'
                        rollTime={1}
                        dieSize={30}
                        outline
                        ref={(dice: DiceRef) => diceRef = dice}
                        rollDone={onRollDone}
                    />
                </CharacterDice> 
            </CharacterInfoDice>
        </CharacterContainer>
    )
}

Character.displayName = 'Character';
export default Character;