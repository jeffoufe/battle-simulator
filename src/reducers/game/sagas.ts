import { ALL_ROLLS_DONE, END_TURN, RESET_TURN, COMPUTE_TURN } from './constants';
import { takeEvery, put, select, delay } from 'redux-saga/effects';

interface EndTurnAction {
    payload: {
        result: number
    }
}

function* endTurn(action: EndTurnAction) {
    const { result } = action.payload;
    yield put({
        type: ALL_ROLLS_DONE,
        payload: { result }
    })
    const playerResult = select((state: any) => state.player.result);
    const monsterResult = select((state: any) => state.monster.result);
    if (playerResult === monsterResult) {
        yield delay(1000);
        yield put({
            type: RESET_TURN
        });
    } else {
        const order = playerResult < monsterResult;
        yield put({
            type: COMPUTE_TURN,
            payload: { order }
        })
        yield delay(1000);
        yield put({
            type: RESET_TURN
        });
    }
}   

export function* watchEndTurn() {
    // @ts-ignore
    yield takeEvery(END_TURN, endTurn);
}