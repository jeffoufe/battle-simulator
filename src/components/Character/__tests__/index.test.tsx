import * as React from 'react';
import { shallow } from 'enzyme';
import Character from '..';
import { INITIAL_STATE } from '../../../reducers/game';
import { monsterRoll, endTurn } from '../../../reducers/game/actions';
import * as redux from "react-redux";

const mockDispatchFn = jest.fn()

describe('Actions', () => {
    beforeEach(() => {
        const useDispatchSpy = jest.spyOn(redux, 'useDispatch'); 
        useDispatchSpy.mockReturnValue(mockDispatchFn);
    })

    it('Player attacks', () => {
        const rollAll = jest.fn();
        const useRefSpy = jest.spyOn(React, 'useRef');
        // @ts-ignore
        useRefSpy.mockReturnValue({ rollAll })

        const useSelectorSpy = jest.spyOn(redux, 'useSelector'); 
        useSelectorSpy.mockReturnValue({ ...INITIAL_STATE.player, isRolling: true });

        jest.spyOn(React, 'useEffect').mockImplementation(f => f());

        const wrapper = shallow(
            <Character order={0} image='image' />
        );
        
        expect(wrapper).toMatchSnapshot();
        expect(rollAll).toHaveBeenCalled();

        const rollDone: (result: number) => void = wrapper.find('CharacterDice').find('ForwardRef').prop('rollDone');
        rollDone(5);
        expect(mockDispatchFn).toHaveBeenCalledWith(monsterRoll(5));
    });

    it('Monster attacks', () => {
        const useSelectorSpy = jest.spyOn(redux, 'useSelector'); 
        useSelectorSpy.mockReturnValue({ ...INITIAL_STATE.monster });

        const wrapper = shallow(
            <Character order={1} image='image' />
        );
        
        expect(wrapper).toMatchSnapshot();
        const rollDone: (result: number) => void = wrapper.find('CharacterDice').find('ForwardRef').prop('rollDone');
        rollDone(3);
        expect(mockDispatchFn).toHaveBeenCalledWith(endTurn(3));
    })
});