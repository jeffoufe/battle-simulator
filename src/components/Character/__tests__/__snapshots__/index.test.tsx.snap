// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Actions Monster attacks 1`] = `
<CharacterContainer
  order={1}
>
  <CharacterName
    order={1}
  >
    Monster
  </CharacterName>
  <CharacterInfoDice
    order={1}
  >
    <CharacterInfo
      order={1}
    >
      <CharacterImage
        src="image"
      />
      <CharacterHealth
        health={100}
        order={1}
      />
    </CharacterInfo>
    <CharacterDice
      order={1}
    >
      <ForwardRef
        defaultRoll={4}
        dieSize={30}
        disableIndividual={false}
        dotColor="#000000"
        faceColor="#FFFFFF"
        margin={15}
        numDice={2}
        outline={true}
        outlineColor="#000000"
        rollDone={[Function]}
        rollTime={1}
        sides={6}
      />
    </CharacterDice>
  </CharacterInfoDice>
</CharacterContainer>
`;

exports[`Actions Player attacks 1`] = `
<CharacterContainer
  order={0}
>
  <CharacterName
    order={0}
  >
    Player
  </CharacterName>
  <CharacterInfoDice
    order={0}
  >
    <CharacterInfo
      order={0}
    >
      <CharacterImage
        src="image"
      />
      <CharacterHealth
        health={100}
        order={0}
      />
    </CharacterInfo>
    <CharacterDice
      order={0}
    >
      <ForwardRef
        defaultRoll={4}
        dieSize={30}
        disableIndividual={false}
        dotColor="#000000"
        faceColor="#FFFFFF"
        margin={15}
        numDice={2}
        outline={true}
        outlineColor="#000000"
        rollDone={[Function]}
        rollTime={1}
        sides={6}
      />
    </CharacterDice>
  </CharacterInfoDice>
</CharacterContainer>
`;
